Lab_1-2-3/lab-1-2-3-sql-window-function.md

```diff
@@ -1168,65 +1215,88 @@ order by categoryid, unitprice desc;
 
+**Zapytania**
+
+_MS SQL_
+
 ```sql
-SELECT p.productid,
-       p.productname,
-       p.unitprice,
-       p.categoryid,
-       (SELECT top 1 productname
-        FROM products
-        WHERE categoryid = p.categoryid
-        ORDER BY unitprice DESC) AS first,
-       (SELECT top 1 productname
-        FROM products
-        WHERE categoryid = p.categoryid and unitprice = p.unitprice
-        ORDER BY unitprice ASC) as last
-FROM products p
-ORDER BY p.categoryid, p.unitprice DESC;
+select p.productid, p.productname, p.unitprice, p.categoryid,
+       (select top 1 p2.productname from product_history p2 where p2.CategoryID=p.CategoryID
+       order by p2.UnitPrice desc) first,
+       (select top 1 p2.productname from product_history p2 where p2.CategoryID=p.CategoryID
+       order by p2.UnitPrice) last
+from product_history p
+order by p.categoryid, p.unitprice desc;
 ```
 
-Dla Postgresa i SQLite zamiast top 1 trzeba użyć limit 1
+_PostgreSQL_
 
 ```sql
-SELECT p.productid,
-       p.productname,
-       p.unitprice,
-       p.categoryid,
-       (SELECT  productname
-        FROM products
-        WHERE categoryid = p.categoryid
-        ORDER BY unitprice DESC limit 1) AS first,
-       (SELECT productname
-        FROM products
-        WHERE categoryid = p.categoryid and unitprice = p.unitprice
-        ORDER BY unitprice ASC limit 1) as last
-FROM products p
-ORDER BY p.categoryid, p.unitprice DESC;
+select p.productid, p.productname, p.unitprice, p.categoryid,
+       (select p2.productname from product_history p2 where p2.CategoryID=p.CategoryID
+       order by p2.UnitPrice desc limit 1) first,
+       (select p2.productname from product_history p2 where p2.CategoryID=p.CategoryID
+       order by p2.UnitPrice limit 1) last
+from product_history p
+order by p.categoryid, p.unitprice desc;
 ```
 
-Bez użycia funkcji okna możemy uzyskać wyniki których potrzebujemy.
+_SQLite_
+
+```sql
+select p.productid, p.productname, p.unitprice, p.categoryid,
+       (select p2.productname from product_history p2 where p2.CategoryID=p.CategoryID
+       order by p2.UnitPrice desc limit 1) first,
+       (select p2.productname from product_history p2 where p2.CategoryID=p.CategoryID
+       order by p2.UnitPrice limit 1) last
+from product_history p
+order by p.categoryid, p.unitprice desc;
+```
+
+> Bez użycia funkcji okna możemy uzyskać wyniki których potrzebujemy.
+
+> W zadaniu tym należało użyć tabeli `product_history`, która posiada 2500 rekordów.
+
+**Czasy wykonania**
+
+| Zapytanie | MS SQL                         | Postgres                          | SQLite                          |
+| --------- | ------------------------------ | --------------------------------- | ------------------------------- |
+| 1         | ![](./img/ex12/mssql1time.png) | ![](./img/ex12/postgres1time.png) | ![](./img/ex12/sqlite1time.png) |
+| 2         | ![](./img/ex12/mssql2time.png) | ![](./img/ex12/postgres2time.png) | ![](./img/ex12/sqlite2time.png) |
```

Comment:
Wyświetlamy 77 wierszy dla zapytania z funkcją okna a 3731959 dla drugiego zapytania dla MS SQL. Chyba oznacza to, że wynik nie jest taki sam?

Reply:
Poprawiony czas